3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/GoogleAnalyticsAdapter.scala

```diff
@@ -0,0 +1,651 @@
+            m + (newName -> translation(value).toValidationNel)
+          }
+          .getOrElse(m)
+    }.sequenceU
+
+  /**
+   * Translates a payload according to a translation table.
+   * @param originalParams original payload in key-value format
+   * @param translationTable mapping between original params and the wanted format
+   * @return a translated params
+   */
+  private def translatePayload(
+    originalParams: Map[String, String],
+    translationTable: Map[String, String]
+  ): Map[String, String] =
+    originalParams.foldLeft(Map.empty[String, String]) {
+      case (m, (fieldName, value)) =>
+        translationTable
+          .get(fieldName)
+          .map(newName => m + (newName -> value))
+          .getOrElse(m)
+    }
+
+  /**
+   * Discovers the contexts in the payload in linear time (size of originalParams).
+   * @param originalParams original payload in key-value format
+   * @param referenceTable map of context schemas and their associated translations
+   * @param fieldToSchemaMap reverse indirection from referenceTable linking fields with the MP
+   * nomenclature to schemas
+   * @return a map containing the discovered contexts keyed by schema
+   */
+  private def buildContexts(
+    originalParams: Map[String, String],
+    referenceTable: Map[String, Map[String, Translation]],
+    fieldToSchemaMap: Map[String, String]
+  ): ValidationNel[String, Map[String, Map[String, FieldType]]] =
+    originalParams.foldLeft(Map.empty[String, Map[String, ValidationNel[String, FieldType]]]) {
+      case (m, (fieldName, value)) =>
+        fieldToSchemaMap.get(fieldName).map { schema =>
+          // this is safe when fieldToSchemaMap is built from referenceTable
+          val (translation, newName) = referenceTable(schema)(fieldName)
+          val trTable = m.getOrElse(schema, Map.empty) +
+            (newName -> translation(value).toValidationNel)
+          m + (schema -> trTable)
+        }
+        .getOrElse(m)
+    }
+    .map { case (k, v) => (k -> v.sequenceU) }
+    .sequenceU
+
+  /**
+   * Builds the contexts containing composite fields in quadratic time
+   * (nr of composite fields * avg size for each of them)
+   * @param originalParams original payload in key-value format
+   * @param referenceTable list of context schemas containing composite fields and their
+   * associated translations
+   * @param nrCompFieldsPerSchema map containing the number of field values in the composite field
+   * name. Used to filter out contexts that might have been erroneously built.
+   * @param indicator indicator used to determine if a key-value has been extracted from the
+   * composite field name
+   * @return a map containing the composite contexts keyed by schema
+   */
+  private def buildCompositeContexts(
+    originalParams: Map[String, String],
+    referenceTable: List[MPData],
+    nrCompFieldsPerSchema: Map[String, Int],
+    indicator: String
+  ): \/[String, List[(String, Map[String, FieldType])]] =
```

Comment:
not Either because I don't want to deal with the not right biased version in 2.11 and I don't like the infix notation of scalaz \/ but I guess that's just preferences

Reply:
I disagree, but that's not a strong preference, so ok.

Reply:
Plus, it might be easier to convert \/ to Either with scalafix if it's not infix once we move off scalaz :D